go-controller/pkg/ovn/policy.go

```diff
@@ -1146,25 +1166,33 @@ func (oc *Controller) handlePeerNamespaceSelector(
 		cache.ResourceEventHandlerFuncs{
 			AddFunc: func(obj interface{}) {
 				namespace := obj.(*kapi.Namespace)
-				// This needs to be a write lock because there's no locking around 'gress policies
-				np.Lock()
-				defer np.Unlock()
-				if !np.deleted {
-					gress.addNamespaceAddressSet(namespace.Name, np.portGroupName)
-				}
+				// Update the ACL ...
+				oc.handlePeerNamespaceSelectorOnUpdate(np, gress, func() bool {
+					// ... on condition that the added address set was not already in the 'gress policy
+					return gress.addNamespaceAddressSet(namespace.Name)
+				})
 			},
 			DeleteFunc: func(obj interface{}) {
 				namespace := obj.(*kapi.Namespace)
-				// This needs to be a write lock
-				np.Lock()
-				defer np.Unlock()
-				if !np.deleted {
-					gress.delNamespaceAddressSet(namespace.Name, np.portGroupName)
-				}
+				// Update the ACL ...
+				oc.handlePeerNamespaceSelectorOnUpdate(np, gress, func() bool {
+					// ... on condition that the removed address set was in the 'gress policy
+					return gress.delNamespaceAddressSet(namespace.Name)
+				})
 			},
 			UpdateFunc: func(oldObj, newObj interface{}) {
 			},
-		}, nil)
+		}, func(i []interface{}) {
+			// This needs to be a write lock because there's no locking around 'gress policies
+			np.Lock()
+			defer np.Unlock()
+			// We load the existing address set into the 'gress policy.
+			// Notice that this will make the AddFunc for this initial
+			// address set a noop.
+			// The ACL must be set explicitly after setting up this handler
+			// for the address set to be considered.
+			gress.addNamespaceAddressSets(i)
```

Comment:
So, lets say that ovn-k is restarting for whatever reason and we had a policy with more that 1 namespace selector.

We would have an ACL with a match like 

`ip4.src == {$<as_nasmepace_1>, $<as_nasmepace_2>, ...} && outputport == @<portgroup>`

Then ovn-k starts and with the changes of this PR we will be updating the ACL, previously we were not doing this. So we setup the namsepace handler for the policy. That handler will be invoking `AddFunc` for each namespace one by one. So for each namespace, first the ACL would be updated to

`ip4.src == {$<as_nasmepace_1>} && outputport == @<portgroup>`

then to 

`ip4.src == {$<as_nasmepace_1>, $<as_nasmepace_2>} && outputport == @<portgroup>`

and finally when the last Addfunc for the laste namespace is processed, updated to it's final form

`ip4.src == {$<as_nasmepace_1>, $<as_nasmepace_2>, ...} && outputport == @<portgroup>`

So until all the namespaces are processed we would not have the ACL where we need it and until that time we would be disallowing traffic that should be allowed and that we had been allowing previously.

What we are doing with the processExisting is that all the namespace selectors are loaded into memory before the first attempt to update the ACL, so the update would be directly to its final desired form.

Hope this clears a bit, let me know.

Reply:
thanks this makes sense. The last part I'm wondering about is this logic will only set the address sets at start up time, so are you banking on this to update the ACL:
https://github.com/ovn-org/ovn-kubernetes/pull/2236/commits/98306249a6d844a60e6c58e431a3b03e37738d2d#diff-cc83e19af1c257d5a09b711d5977d8f8c20beb34b7b5d3eb37b2f2c53ded1bf7R1065

because it looks like since there is no address set update thats going to happen for subsequent informer namespace add's for previous sync'ed objects, your doUpdate would return false for each add event, and the ACL wouldnt get updated otherwise.

Reply:
Yes, that's right. We rely on that to make the first update to the ACL with the `processExistingItems`. Then the `AddFunc` of the handler will be called for each existing namespace selectors but indeed doUpdate returns `false` because those were already loaded into memory so no update happens. And then for further namespace selectors added updating the policy, `doUpdate` would be `true`and would trigger an update of the ACL.